File: problem solving patterns/anagrams.rb
def valid_anagram(str1, str2)
  arr1 = str1.split('')
  arr2 = str2.split('')
  lookup = Hash.new(0)

  arr1.each do |l|
    lookup[l] += 1
  end

  for l in arr2
    return false if lookup[l] <= 0;
    lookup[l] -= 1
  end
  
  true
end

puts valid_anagram('', '') # => true
puts valid_anagram('aaz', 'zza') # => false
puts valid_anagram('anagram', 'nagaram') # => true
puts valid_anagram("rat","car") # => false
puts valid_anagram('awesome', 'awesom') # => false
puts valid_anagram('qwerty', 'qeywrt') # => true
puts valid_anagram('texttwisttime', 'timetwisttext') # => true

File: problem solving patterns/anagrams.js
/* Given two strings, write a function to determine if the second string is an anagram of the first. An anagram is a word, phrase, or name formed by rearranging the letters of another, such as cinema, formed from iceman */


const validAnagram = (str1, str2) => {
  // This should return a boolean

  if (str1.length !== str2.length) return false;

  const frequencyCounter = {}
  const wordsArr = str1.split('')

  wordsArr.forEach(l => {
    frequencyCounter[l] = (frequencyCounter[l] || 0) + 1
  })
  
  for (let l of str2) {
    if(!frequencyCounter[l])  {
      return false
    } else {
      frequencyCounter[l] --
    }
  }
  return true
}

console.log(validAnagram('', ''))// true
console.log(validAnagram('aaz', 'zza')) // false
console.log(validAnagram('anagram', 'nagaram')) // true
console.log(validAnagram("rat","car")) // false
console.log(validAnagram('awesome', 'awesom')) // false
console.log(validAnagram('qwerty', 'qeywrt')) // true
console.log(validAnagram('texttwisttime', 'timetwisttext')) // true
